### PR TITLE
[5.1] Fixed A Load Of Issues With The Request Class

### PR DESCRIPTION
**Pull request by GrahamCampbell**
Added Robust REQUEST Content Type Parsing

This also fixes a small issues with the accept header, and abstracts some code both use.

Note that the accept header parsing code and the request content type parsing code call matchesType with paramaters in the reverse order.

### CODE REVIEW

**Comment by wells** [comment]
:+1:

**Comment by JerseyMilker** [comment]
I like this @GrahamCampbell.  It doesn't break anything for 5.1, and it isn't overly permissive like the PR I submitted for `isJson()`.  Yet it still allows for jsonapi.org nuts.  I say 'nuts', because free speech.

File: src/Illuminate/Http/Request.php [hunk 1]
```diff
@@ -566,13 +566,48 @@ protected function getInputSource()
...
 
     /**
+     * Do the two content types match.
+     *
+     * If the first type is the same as the second type, or is a subset, then
+     * we're returning true, otherwise, false.
+     *
+     * @return bool
+     */
+    public function matchesType($actual, $type)
```
Comment (franzliedke) [inline src/Illuminate/Http/Request.php]: Do we need this as a public method on the request class?

Comment (GrahamCampbell) [inline src/Illuminate/Http/Request.php]: Yes. I think it's a useful method for people to use.

**Comment by franzliedke** [comment]
:-1:, as this is breaking.

Even if you say that `text/json` is not a valid JSON request, it has been recognized as such before. And judging by previous decisions, this should not be considered a "bug" worth breaking backwards compatibility for.

**Comment by GrahamCampbell** [comment]
> Even if you say that text/json is not a valid JSON request, it has been recognized as such before.

I think that's a bug. It's not our fault if people were using something that was broken.

**Comment by franzliedke** [comment]
The famous "You're using it wrong".

**Comment by GrahamCampbell** [comment]
@franzliedke All bug fixes are BC breaks...

**Comment by GrahamCampbell** [comment]
They used to work in a broken way, and now the broken way doesn't work. :P

**Comment by JerseyMilker** [comment]
`application/json` is the only valid mimetype for a JSON file.  It's a shame jsonapi.org and similar enforce a non-standard content-type for request header.  After all, jsonapi.org is still enforcing valid JSON syntax, it's just opinionated on how that JSON is structured for API use cases (schema convention, key naming, nesting, etc.).

All that said, I still agree with this PR.  Whether I agree with jsonapi.org's implementation or not, if someone chooses to send a JSON request using jsonapi.org convention, `isJson()` method should still return true, because it will be JSON data coming in.

Regarding `text/json`, unless another API convention similar to jsonapi.org uses this, I don't see reason to account for it.  I think `isJson()` should be as strict as possible, while allowing for API conventions like jsonapi.org, HAL, etc.

**Comment by joecohens** [comment]
:+1:

**Comment by JerseyMilker** [comment]
@GrahamCampbell Should you account for this:
`Content-type: application/json; charset=utf-8`

I think the charset is pretty redundant, but there's chance some people could pass other valid info like this into content-type.  It's an edge case, but maybe you should explode on `;` and loop through array?

**Comment by GrahamCampbell** [comment]
> Should you account for this:

We do account for it already. Symfony deals with that before we even process anything.

**Comment by GrahamCampbell** [comment]
Oh, right, I see what You mean now, with the content type. I forgot about that because symfony deals with it in the request header for us. :)

**Comment by JerseyMilker** [comment]
Ah I see.  Write a test for this anyway?
`Content-type: application/json; charset=utf-8`

**Comment by GrahamCampbell** [comment]
Yeh, I'll add some tests. :)

**Comment by JerseyMilker** [comment]
Awesome :+1:

**Comment by taylorotwell** [comment]
I feel like we're making a lot of changes to request to satisfy one user, and not with any clear description of any problems.

**Comment by JerseyMilker** [comment]
@taylorotwell I described the issue in more detail at #9325.  My PR wasn't necessarily a good solution, so @GrahamCampbell  offered his take on the issue here.

If a 3rd party is sending JSON requests using jsonapi.org spec, or something like HAL, Siren, etc., then Content-Type might vary from `application/json`.  Many of these JSON API "standards" encourage valid JSON syntax, but are instead opinionated on how JSON is structured for API (key naming, nesting practices, structural things relating to API use cases).  Yet, the current `isJson()` check returns false if an API uses any of these conventions because of the content-type they enforce.

As per jsonapi.org docs: ([Source](http://jsonapi.org/format/))
>Clients MUST send all JSON API data in request documents with the header Content-Type: **application/vnd.api+json** without any media type parameters.

Some Content-Type possibilities:
```
application/vnd.api+json    // jsonapi.org
application/hal+json        // HAL
application/vnd.siren+json  // Siren
```

>to satisfy one user

I see your point, but I think some of these are gaining more steam than we think.  Whether we agree on jsonapi.org convention or not, it's still valid JSON syntax.  I think `isJson()` should reflect that.  In a way, it's like saying "PSR-2 isn't real PHP".

**Comment by taylorotwell** [comment]
Thanks for providing more real description. Can someone elaborate on the
breaking change implications? Are there any breaking changes? If so, what
are they?

**Comment by JerseyMilker** [comment]
No problem.  I'm not sure I'd consider the changes "breaking".  As mentioned in my last comment, right now the `isJson()` method rejects valid JSON if the content-type is associated with a JSON convention like jsonapi.org.  This is a fix for that.  The `isJson()` method is asking, "Is this request JSON?".  Now it will be returning a more accurate boolean.

This PR also ensures charset is properly handled when passed into content-type header.  Passing charset like this is valid according to w3.org ([source](http://www.w3.org/International/O-HTTP-charset)):
`Content-Type: application/json; charset=utf-8`

I'm not sure if any of these changes are breaking, unless someone is wrongly relying on the broken functionality of `isJson()` returning false for something like `application/vnd.api+json`.

**Comment by franzliedke** [comment]
This is breaking, because `isJson()` would not recognize content types like `text/json` anymore.

Quoting @JerseyMilker:

> I see your point, but I think some of these are gaining more steam than we think. Whether we agree on jsonapi.org convention or not, it's still valid JSON syntax. I think `isJson()` should reflect that. In a way, it's like saying "PSR-2 isn't real PHP".

I agree, that's why `text/json` should be recognized, too.

The other changes are fine and will probably determine the best practice for the future. We can change the behaviour of `isJson` in the next major release.

**Comment by JerseyMilker** [comment]
@franzliedke, I never got a reply earlier for this:

>Regarding text/json, unless another API convention similar to jsonapi.org uses this, I don't see reason to account for it. I think isJson() should be as strict as possible, while allowing for API conventions used by jsonapi.org, HAL, Siren, etc. @franzliedke Why might someone use text/json?

Could you comment on this?  A bit of google searching revealed that `text/json` is not an official mime type for JSON.  I'm still unclear why anyone would use it over `application/json`?

**Comment by franzliedke** [comment]
It's not in the spec, but I've seen it being used in the wild. There's no reason to break this for people on a minor release.

**Comment by GrahamCampbell** [comment]
@franzliedke We can always revert that part of this.

**Comment by franzliedke** [comment]
Exactly my point.

**Comment by GrahamCampbell** [comment]
Ping @taylorotwell.

**Comment by taylorotwell** [comment]
Yeah, revert that part of it.

**Comment by adamwathan** [comment]
+1 for allowing jsonapi.org compliant requests to pass the `isJson` check. I really expect that to pick up a lot of steam now that it's 1.0 because of the people involved (lots of important Ember and Rails core team members).

**Pull request closed without merging**